File: ios/NorthAx/Engine/ZoneMath.swift
import Foundation

// Converts coach-emitted zone tokens (Z1..Z5) into concrete numeric ranges
// using the athlete's thresholds. Pure / no SwiftUI. The graph uses this to
// draw a real numeric Y-axis; when the needed threshold is missing every
// function returns nil and the graph falls back to zone-only rendering.

enum ZoneMode { case hr, power, pace }

/// Numeric bounds for a zone. Units depend on mode: bpm (hr), watts (power),
/// seconds-per-unit (pace, where LOWER = faster). A nil bound is open-ended.
struct ZoneRange {
    let lower: Double?
    let upper: Double?
}

enum ZoneMath {

    // % of FTP. Z5 is open-topped; cap display at 120%.
    private static let powerBounds: [Int: (Double, Double?)] = [
        1: (0.40, 0.55), 2: (0.56, 0.75), 3: (0.76, 0.90),
        4: (0.91, 1.05), 5: (1.06, 1.20),
    ]
    // % of LTHR. Z5 open-topped; cap ~110%.
    private static let hrBounds: [Int: (Double, Double?)] = [
        1: (0.65, 0.81), 2: (0.81, 0.89), 3: (0.90, 0.93),
        4: (0.94, 0.99), 5: (1.00, 1.10),
    ]
    // % of threshold SPEED (pace is inverse). Z5 open-topped; cap ~108%.
    private static let paceSpeedBounds: [Int: (Double, Double?)] = [
        1: (0.78, 0.84), 2: (0.84, 0.91), 3: (0.91, 0.96),
        4: (0.96, 1.02), 5: (1.02, 1.08),
    ]

    /// Numeric range for a zone, or nil if the relevant threshold is absent.
    static func range(zone: Int, mode: ZoneMode, sport: TrainingDomain,
                      thresholds: AthleteThresholds) -> ZoneRange? {
        guard (1...5).contains(zone) else { return nil }
        switch mode {
        case .power:
            guard let ftp = thresholds.ftpWatts, let b = powerBounds[zone] else { return nil }
            return ZoneRange(lower: Double(ftp) * b.0, upper: b.1.map { Double(ftp) * $0 })
        case .hr:
            guard let lthr = lthr(thresholds), let b = hrBounds[zone] else { return nil }
            return ZoneRange(lower: lthr * b.0, upper: b.1.map { lthr * $0 })
        case .pace:
            guard let thr = thresholdPaceSeconds(sport: sport, thresholds: thresholds),
                  let b = paceSpeedBounds[zone] else { return nil }
            // Higher speed factor => faster => fewer seconds. Lower seconds bound
            // corresponds to the FASTER (upper) speed factor.
            let lowerSec = b.1.map { Double(thr) / $0 }       // fastest end (may be open)
            let upperSec = Double(thr) / b.0                   // slowest end
            return ZoneRange(lower: lowerSec, upper: upperSec)
        }
    }

    /// Representative value for plotting the segment height.
    static func midpoint(zone: Int, mode: ZoneMode, sport: TrainingDomain,
                         thresholds: AthleteThresholds) -> Double? {
        guard let r = range(zone: zone, mode: mode, sport: sport, thresholds: thresholds)
        else { return nil }
        switch (r.lower, r.upper) {
        case let (l?, u?): return (l + u) / 2
        case let (l?, nil): return l       // open-topped: anchor at the floor
        case let (nil, u?): return u
        default: return nil
        }
    }

    /// Human-readable range string with units.
    static func format(_ range: ZoneRange, mode: ZoneMode,
                       sport: TrainingDomain = .running,
                       paceUnit: PaceUnit = .km) -> String {
        switch mode {
        case .hr:
            return "\(intStr(range.lower))–\(intStr(range.upper)) bpm"
        case .power:
            return "\(intStr(range.lower))–\(intStr(range.upper)) W"
        case .pace:
            let suffix = sport == .swimming ? "/100m" : (paceUnit == .mile ? "/mi" : "/km")
            // Faster (lower seconds) shown first.
            return "\(paceStr(range.lower))–\(paceStr(range.upper))\(suffix)"
        }
    }

    // MARK: - Helpers

    /// LTHR: prefer measured threshold HR, else estimate from max HR (≈0.92·max).
    private static func lthr(_ t: AthleteThresholds) -> Double? {
        if let hr = t.thresholdHr { return Double(hr) }
        if let mx = t.maxHr { return Double(mx) * 0.92 }
        return nil
    }

    private static func thresholdPaceSeconds(sport: TrainingDomain,
                                             thresholds: AthleteThresholds) -> Int? {
        sport == .swimming ? thresholds.swimThresholdPaceSecPer100m
                           : thresholds.runThresholdPaceSecPerKm
    }

    private static func intStr(_ v: Double?) -> String {
        guard let v else { return "–" }
        return String(Int(v.rounded()))
    }

    private static func paceStr(_ seconds: Double?) -> String {
        guard let seconds else { return "–" }
        let s = Int(seconds.rounded())
        return "\(s / 60):" + String(format: "%02d", s % 60)
    }
}

File: android/app/src/main/java/app/northax/domain/engine/ZoneMath.kt
package app.northax.domain.engine

import app.northax.domain.model.AthleteThresholds
import app.northax.domain.model.PaceUnit
import app.northax.domain.model.TrainingDomain
import kotlin.math.roundToInt

// Converts coach-emitted zone tokens (Z1..Z5) into concrete numeric ranges
// using the athlete's thresholds. Pure / no UI. The graph uses this to draw a
// real numeric Y-axis; when the needed threshold is missing every function
// returns null and the graph falls back to zone-only rendering.

enum class ZoneMode { Hr, Power, Pace }

/**
 * Numeric bounds for a zone. Units depend on mode: bpm (hr), watts (power),
 * seconds-per-unit (pace, where LOWER = faster). A null bound is open-ended.
 */
data class ZoneRange(val lower: Double?, val upper: Double?)

object ZoneMath {

    // % of FTP. Z5 is open-topped; cap display at 120%.
    private val powerBounds: Map<Int, Pair<Double, Double?>> = mapOf(
        1 to (0.40 to 0.55), 2 to (0.56 to 0.75), 3 to (0.76 to 0.90),
        4 to (0.91 to 1.05), 5 to (1.06 to 1.20),
    )

    // % of LTHR. Z5 open-topped; cap ~110%.
    private val hrBounds: Map<Int, Pair<Double, Double?>> = mapOf(
        1 to (0.65 to 0.81), 2 to (0.81 to 0.89), 3 to (0.90 to 0.93),
        4 to (0.94 to 0.99), 5 to (1.00 to 1.10),
    )

    // % of threshold SPEED (pace is inverse). Z5 open-topped; cap ~108%.
    private val paceSpeedBounds: Map<Int, Pair<Double, Double?>> = mapOf(
        1 to (0.78 to 0.84), 2 to (0.84 to 0.91), 3 to (0.91 to 0.96),
        4 to (0.96 to 1.02), 5 to (1.02 to 1.08),
    )

    /** Numeric range for a zone, or null if the relevant threshold is absent. */
    fun range(zone: Int, mode: ZoneMode, sport: TrainingDomain, thresholds: AthleteThresholds): ZoneRange? {
        if (zone !in 1..5) return null
        return when (mode) {
            ZoneMode.Power -> {
                val ftp = thresholds.ftpWatts ?: return null
                val b = powerBounds[zone] ?: return null
                ZoneRange(ftp * b.first, b.second?.let { ftp * it })
            }
            ZoneMode.Hr -> {
                val lthr = lthr(thresholds) ?: return null
                val b = hrBounds[zone] ?: return null
                ZoneRange(lthr * b.first, b.second?.let { lthr * it })
            }
            ZoneMode.Pace -> {
                val thr = thresholdPaceSeconds(sport, thresholds) ?: return null
                val b = paceSpeedBounds[zone] ?: return null
                // Higher speed factor => faster => fewer seconds. Lower seconds
                // bound corresponds to the FASTER (upper) speed factor.
                val lowerSec = b.second?.let { thr / it } // fastest end (may be open)
                val upperSec = thr / b.first              // slowest end
                ZoneRange(lowerSec, upperSec)
            }
        }
    }

    /** Representative value for plotting the segment height. */
    fun midpoint(zone: Int, mode: ZoneMode, sport: TrainingDomain, thresholds: AthleteThresholds): Double? {
        val r = range(zone, mode, sport, thresholds) ?: return null
        return when {
            r.lower != null && r.upper != null -> (r.lower + r.upper) / 2
            r.lower != null -> r.lower // open-topped: anchor at the floor
            r.upper != null -> r.upper
            else -> null
        }
    }

    /** Human-readable range string with units. */
    fun format(
        range: ZoneRange,
        mode: ZoneMode,
        sport: TrainingDomain = TrainingDomain.Running,
        paceUnit: PaceUnit = PaceUnit.Km,
    ): String = when (mode) {
        ZoneMode.Hr -> "${intStr(range.lower)}–${intStr(range.upper)} bpm"
        ZoneMode.Power -> "${intStr(range.lower)}–${intStr(range.upper)} W"
        ZoneMode.Pace -> {
            val suffix = if (sport == TrainingDomain.Swimming) "/100m"
            else if (paceUnit == PaceUnit.Mile) "/mi" else "/km"
            // Faster (lower seconds) shown first.
            "${paceStr(range.lower)}–${paceStr(range.upper)}$suffix"
        }
    }

    // MARK: - Helpers

    /** LTHR: prefer measured threshold HR, else estimate from max HR (≈0.92·max). */
    private fun lthr(t: AthleteThresholds): Double? {
        t.thresholdHr?.let { return it.toDouble() }
        t.maxHr?.let { return it * 0.92 }
        return null
    }

    private fun thresholdPaceSeconds(sport: TrainingDomain, thresholds: AthleteThresholds): Double? =
        (if (sport == TrainingDomain.Swimming) thresholds.swimThresholdPaceSecPer100m
        else thresholds.runThresholdPaceSecPerKm)?.toDouble()

    private fun intStr(v: Double?): String = v?.roundToInt()?.toString() ?: "–"

    private fun paceStr(seconds: Double?): String {
        val s = seconds?.roundToInt() ?: return "–"
        return "${s / 60}:" + "%02d".format(s % 60)
    }
}
